src/mapping/query/function/format_url.rs

```diff
@@ -0,0 +1,235 @@
+const USERNAME: &str = "username";
+const PASSWORD: &str = "password";
+const HOST: &str = "host";
+const PORT: &str = "port";
+const PATH: &str = "path";
+const FRAGMENT: &str = "fragment";
+const QUERY: &str = "query";
+
+#[derive(Debug)]
+pub(in crate::mapping) struct FormatUrlFn {
+    query: Box<dyn Function>,
+}
+
+impl FormatUrlFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(query: Box<dyn Function>) -> Self {
+        Self { query }
+    }
+}
+
+impl Function for FormatUrlFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let components = required!(ctx, self.query, Value::Map(v) => v);
+
+        components_to_url(components).map(|url| url.into_string().into())
+    }
+
+    fn parameters() -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            accepts: |v| matches!(v, Value::Map(_)),
+            required: true,
+        }]
+    }
+}
+
+impl TryFrom<ArgumentList> for FormatUrlFn {
+    type Error = String;
+
+    fn try_from(mut arguments: ArgumentList) -> Result<Self> {
+        let query = arguments.required("value")?;
+
+        Ok(Self { query })
+    }
+}
+
+fn components_to_url(components: BTreeMap<String, Value>) -> Result<Url> {
+    let scheme = components
+        .get(SCHEME)
+        .map(|v| string(SCHEME, v))
+        .transpose()?
+        .ok_or_else(|| missing(SCHEME))?;
+
+    let username = components
+        .get(USERNAME)
+        .map(|v| string(USERNAME, v))
+        .transpose()?
+        .unwrap_or_else(|| "".to_owned());
+
+    let password = components
+        .get(PASSWORD)
+        .and_then(Value::as_option)
+        .map(|v| string(PASSWORD, v))
+        .transpose()?;
+
+    let host = components
+        .get(HOST)
+        .and_then(Value::as_option)
+        .map(|v| string(HOST, v))
+        .transpose()?;
+
+    let port = components
+        .get(PORT)
+        .and_then(Value::as_option)
+        .map(|v| {
+            v.as_i64()
+                .map(|i| i as u16)
+                .ok_or_else(|| invalid(PORT, "map", v.kind()))
+        })
+        .transpose()?;
+
+    let path = components
+        .get(PATH)
+        .map(|v| string(PATH, v))
+        .transpose()?
+        .unwrap_or_else(|| "".to_owned());
+
+    let fragment = components
+        .get(FRAGMENT)
+        .and_then(Value::as_option)
+        .map(|v| string(FRAGMENT, v))
+        .transpose()?;
+
+    let query = components
+        .get(QUERY)
+        .map(|v| v.as_map().ok_or_else(|| invalid(QUERY, "map", v.kind())))
+        .transpose()?
+        .map(|v| {
+            v.iter()
+                .map(|(k, v)| (k, v.to_string_lossy()))
+                .collect::<Vec<_>>()
+        })
+        .unwrap_or_default();
+
+    // This is a really awkward way to create a valid URL from components, but
+    // it doesn't appear the `Url` API offers any better way to handle this.
+    let mut url = Url::parse("https://vector.dev").expect("valid");
```

Comment:
I'm really not a fan of this, but I didn't see any other option, aside from formatting manually using `format!`, but that's rather tricky to get correct, given the esoteric rules w.r.t. valid URL parsing.

I looked into using [`http::Uri`](https://docs.rs/http/0.2.1/http/uri/struct.Uri.html), which we already use in Vector, as it provides a `from_parts` function, but that type (among other issues) [doesn't support parsing the _fragment_ part of the URL](https://github.com/hyperium/http/issues/127).

Anyone have a better alternative to the current approach?

Reply:
It feels clunky, but I can't think of a better way. 

It might be worth raising an issue with `rust-url` to see if we could have some kind of `UrlBuilder` object that would allow you to build up a Url without having to parse one first? I can't imagine we are the only people who would find that useful..

Reply:
Hmm, yeah, I'm :+1: on filing an issue for a builder with that crate.